### PR TITLE
Support generic OIDC auth with ID tokens

### LINKED ISSUE #126: Support generic OIDC implementation
Add an OIDC implementation where the id token can be provided directly. This will enable the use of Terraform with OIDC from Kubernetes, GitLab and others.

The current OIDC implementation (#91) only supports the GitHub Actions way of providing the id token via a HTTP call. 

See also:
- https://github.com/hashicorp/terraform-provider-azurerm/issues/16901
- https://github.com/manicminer/hamilton/pull/166

### PR DESCRIPTION
**Pull request by derkoe**
Closes #126

### CODE REVIEW

File: authentication/auth_method_oidc.go [hunk 1]
```diff
@@ -32,7 +35,7 @@ func (a oidcAuth) build(b Builder) (authMethod, error) {
 }
 
 func (a oidcAuth) isApplicable(b Builder) bool {
-	return b.SupportsOIDCAuth && b.IDTokenRequestURL != "" && b.IDTokenRequestToken != "" && b.UseMicrosoftGraph
+	return b.SupportsOIDCAuth && (b.IDToken != "" || (b.IDTokenRequestURL != "" && b.IDTokenRequestToken != "")) && b.UseMicrosoftGraph
```
Comment (manicminer) [inline authentication/auth_method_oidc.go:38]: Just a small suggestion to make this slightly more readable

```suggestion
	return b.SupportsOIDCAuth && b.UseMicrosoftGraph && (b.IDToken != "" || (b.IDTokenRequestURL != "" && b.IDTokenRequestToken != ""))
```

**Review by manicminer — APPROVED** [review]
Hi @derkoe, thanks very much for this contribution. This feels like an elegant way to add support for a supplied ID token and I'm happy to accept this change :)

**Comment by derkoe** [comment]
Thanks @manicminer! After the next release I will create the PR in the AzureRM provider - current WIP is here https://github.com/hashicorp/terraform-provider-azurerm/compare/main...derkoe:terraform-provider-azurerm:generic-oidc-auth

**Comment by manicminer** [comment]
Thanks @derkoe, this has been included in v0.40.0

**Pull request merged**